src/passkey_child/passkey_child_assert.c

```diff
@@ -0,0 +1,210 @@
+    }
+
+done:
+    return ret;
+}
+
+errno_t
+prepare_assert(const struct passkey_data *data, int index,
+               fido_assert_t *_assert)
+{
+    TALLOC_CTX *tmp_ctx = NULL;
+    unsigned char *key_handle;
+    size_t key_handle_len;
+    errno_t ret;
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = fido_assert_set_rp(_assert, data->domain);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "fido_assert_set_rp failed [%d]: %s.\n", ret, fido_strerr(ret));
+        goto done;
+    }
+
+    key_handle = sss_base64_decode(tmp_ctx, data->key_handle_list[index],
+                                   &key_handle_len);
+    if (key_handle == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to decode key handle.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = fido_assert_allow_cred(_assert, key_handle, key_handle_len);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "fido_assert_allow_cred failed [%d]: %s.\n", ret, fido_strerr(ret));
+        goto done;
+    }
+
+    ret = set_assert_options(FIDO_OPT_FALSE, FIDO_OPT_OMIT, _assert);
+    if (ret != FIDO_OK) {
+        goto done;
+    }
+
+    ret = set_assert_client_data_hash(_assert);
+    if (ret != EOK) {
+        goto done;
+    }
+
+done:
+    talloc_free(tmp_ctx);
+
+    return ret;
+}
+
+errno_t
+reset_public_key(struct pk_data_t *_pk_data)
+{
+    if (_pk_data->type == COSE_ES256) {
+        es256_pk_free((es256_pk_t **) &_pk_data->public_key);
+    } else if (_pk_data->type == COSE_RS256) {
+        rs256_pk_free((rs256_pk_t **) &_pk_data->public_key);
+    } else if (_pk_data->type == COSE_EDDSA) {
+        eddsa_pk_free((eddsa_pk_t **) &_pk_data->public_key);
+    }
+
+    return EOK;
+}
+
+errno_t
+request_assert(struct passkey_data *data, fido_dev_t *dev,
+               fido_assert_t *_assert)
+{
+    bool has_pin;
+    bool has_uv;
+    errno_t ret;
+
+    has_uv = fido_dev_has_uv(dev);
+    if (has_uv == true && data->user_verification != FIDO_OPT_FALSE) {
+        ret = fido_dev_get_assert(dev, _assert, NULL);
+        if (ret != FIDO_OK) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "fido_dev_get_assert failed [%d]: %s. "
+                  "Falling back to PIN authentication.\n",
+                  ret, fido_strerr(ret));
+        } else {
+            DEBUG(SSSDBG_TRACE_FUNC, "fido_dev_get_assert succeeded.\n");
+            goto done;
+        }
+    }
+
+    has_pin = fido_dev_has_pin(dev);
+    if (has_pin == true && data->user_verification != FIDO_OPT_FALSE
+        && data->pin == NULL) {
+        ret = ERR_INPUT_PARSE;
+        ERROR("PIN required.\n");
+        goto done;
+    }
+
+    ret = fido_dev_get_assert(dev, _assert, data->pin);
+    if (ret != FIDO_OK) {
+        DEBUG(SSSDBG_OP_FAILURE, "fido_dev_get_assert failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
```

Comment:
>     * The exact same assertion request will be sent to the authenticator twice if `has_uv=true`, `has_pin=false`, `data->user_verification != FIDO_OPT_FALSE`, and the first `fido_dev_get_assert()` fails for some reason. While this is not necessarily a problem, it might be unexpected.

You are right so I added an extra check for it and if there isn't any PIN to fallback then the process fails.

>     * If `has_uv=false`, `has_pin=true`, and `data->user_verification=FIDO_OPT_OMIT`, then the call to `fido_assert_verify()` will not check whether the UV flag was set in the authenticator data. If you want it to be checked, `fido_assert_set_uv(assert, FIDO_OPT_TRUE)` must be called at some point after the second `fido_dev_get_assert()`.
> 
> 
> In general, UV handling might be more clear/concise if it's all done in the same place as the assertion request.

I added a new call to `fido_assert_set_uv()` after calling `fido_dev_get_assert()` to take it into account.